### PR TITLE
websocket: Allow UTF-8 header values

### LINKED ISSUE #12192: Cockpit can't handle HTTP headers that have unicode characters
Cockpit version: 196
OS: Ubuntu 18.04.2 LTS
Page: N/A

Cockpit can't handle unicode characters in HTTP headers. This probably won't ever be hit if user connects directly to cockpit.socket or via simple reverse proxy. 

However I use caddy reverse proxy with JWT used as preauth. One of identifty providers, Github, returns full name including unicode characters in Token-Claim-Name

This causes message "received invalid HTTP request headers" in logs, when debug is enabled additional line appeared - "received invalid header" which pinpoints issue to `is_valid_line` function: https://github.com/cockpit-project/cockpit/blob/899ef8edb8b55b06e9647250842ccd0fb92f9944/src/websocket/websocket.c#L147

Cockpit config untouched. Caddy config:
```
(auth) {
  jwt {
    path /
    allow sub danielskowronski
  }
}
cockpit.example.com {
  import auth
  proxy / http://127.0.0.1:9090 {
    insecure_skip_verify
    transparent
    websocket
    header_upstream Token-Claim-Name _
  }
  log / /var/log/caddy-cockpit.log
}
```

Steps to reproduce
1. Send any header containing unicode to Cockpit (eg. from web browser console, curl)
2. It'll return 500 code

Temporal workaround

In caddy one can overwrite any header sent to backend application, so it's enoung to add `header_upstream Token-Claim-Name _` in `proxy` section of Caddy.

### PR DESCRIPTION
**Pull request by martinpitt**
The headers we look at should all be fine with plain ASCII values. But
it can happen that reverse proxies add extra headers with non-ASCII
values. UTF-8 should be safe with our code also for the headers Cockpit
looks at by itself, so allow these.

Fixes #12192

 - [x] audit all code that looks at headers for whether they assume ASCIIness 
 - [x] enforce ASCIIness of `Sec-WebSocket-Protocol` header
 - [x] enforce ASCIIness of `Authorization` header

### CODE REVIEW

**Comment by martinpitt** [comment]
> Which HTTP specification states that UTF-8 (particularly above characters above 0x80) are permitted in HTTP headers? We should explicitly include a reference to that RFC in the commit message and code comments.

I mentioned that in the associated issue. The [HTTP spec](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2) does not restrict the values or prescribes an encoding; even binary data is allowed as long as it's terminated with CRLF.

> We seem to also be dropping some validation in `is_valid_line()`. Are we explicitly allowing naked low characters? What about `\r` or `\v`.

This  isn't what this patch does. `is_valid_line()` doesn't change -- we use it in a lot of places. This patch only stops using that in  favor of `g_utf8_validate()` for validating the value of a header.

At that point, the line-by-line parsing already happened; `value` is just the thing between the key's `:` and the next newline.

> I'm worried that this opens up security problems through Cockpit. Assumptions that other code makes based on the fact that we only allowed ASCII in.

Right, that's my main concern. I already looked at a few places that use headers, but not all of them. I first wanted to get your general opinion on whether this is a sensible step.

**Comment by martinpitt** [comment]
I did an audit of headers we look at wrt. UTF-8 safety, using this command:

    git grep -E 'lookup.*headers|_web_socket_util_header_contains' | grep -v test-

Header | used for/audit
---------  | ---------
Accept-Language  | string parsing; could hiccup with UTF-8 if e. g. the second byte of a multi-byte character is `:` -- at worst, page is not translated, which is not a security issue
Host | redirection to https (safe enough with UTF-8 chars, the target host might just not exist); comparison to `localhost` constant (safe); concatenation with URL into "origin" for `cockpit_web_response_security_policy()`, where it's copied verbatim into a larger string in `string_inject_origin()`, safe
Content-Length | parsed with `g_ascii_strtoull()` with proper error handling
Transfer-Encoding | only compared to constant `chunked`
Connection | only compared to constants `keep-alive` or `upgrade`
X-Forwarded-{Proto,Host} | concatenated into `Access-Control-Allow-Origin:` header URL, safe
ProtocolHeader option in cockpit.conf | only compared to constants `http` and `https`
Upgrade | only compared to constant `websocket`
X-Cockpit-Pkg-Checksum | not interpreted, just stored verbatim
Pragma | only tested for substring "nocache", safe
If-None-Match | tested for identity to ETag, safe
X-SSH-Connect-Unknown-Hosts | only compared to constant `yes`
Authorization | calculates type prefix up to `:` or ` `; safe, may result in failed auth; type prefix is only compared to constants; auth value is either being passed on as opaque value, or dissected in `cockpit_authorize_parse_basic()` (looks safe, but not 100% sure) :warning:
X-Authorize | only tested for substring "password", safe
Sec-WebSocket-Protocol | Split by `, `, might catch multi-byte UTF-8 sequences :warning: 
Sec-WebSocket-Key | passed to `validate_rfc6455_websocket_key()` which gets base64-decoded; UTF-8 is invalid base64
Origin | used with `g_ascii_strcasecmp()`, which is not safe for arbitrary binary data, but for UTF-8

So in summary, most are safe except for `Authorization` and `Sec-WebSocket-Protocol`; I can't directly see a vulnerability with UTF-8, but just to be on the safe side, let's enforce ASCIIness for these two. I'll rework the PR.

**Comment by mvollmer** [comment]
>  the second byte of a multi-byte character is `:`

This can not happen with UTF-8, and that's why everybody loves it.  string parsing code that looks for ASCII characters will still work with UTF-8.

**Comment by martinpitt** [comment]
Reworked with the hardening discussed above.

**Comment by martinpitt** [comment]
@mvollmer: Hah, so much the better :grin: Nevertheless, I think the two extra checks can't hurt, they just keep the current behaviour.
